DDG4/edm4hep/EDM4hepFileReader.cpp

```diff
@@ -0,0 +1,328 @@
+  /// extension
+  template <class T=podio::GenericParameters> void RunParameters::ingestParameters(T const& source) {
+    for(auto const& key: source.template getKeys<int>()) {
+      m_intValues[key] = source.template get<std::vector<int>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<float>()) {
+      m_fltValues[key] = source.template get<std::vector<float>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<double>()) {
+      m_dblValues[key] = source.template get<std::vector<double>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<std::string>()) {
+      m_strValues[key] = source.template get<std::vector<std::string>>(key).value();
+    }
+  }
+
+  template <class T=podio::GenericParameters> void FileParameters::ingestParameters(T const& source) {
+    for(auto const& key: source.template getKeys<int>()) {
+      m_intValues[key] = source.template get<std::vector<int>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<float>()) {
+      m_fltValues[key] = source.template get<std::vector<float>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<double>()) {
+      m_dblValues[key] = source.template get<std::vector<double>>(key).value();
+    }
+    for(auto const& key: source.template getKeys<std::string>()) {
+      m_strValues[key] = source.template get<std::vector<std::string>>(key).value();
+    }
+  }
+
+  /// Class to read EDM4hep files
+  /**
+   *  \version 1.0
+   *  \ingroup DD4HEP_SIMULATION
+   */
+  class EDM4hepFileReader : public Geant4EventReader {
+  protected:
+    /// Reference to reader object
+    podio::Reader m_reader;
+    /// Name of the MCParticle collection to read
+    std::string m_collectionName;
+    /// Name of the EventHeader collection to read
+    std::string m_eventHeaderCollectionName;
+
+  public:
+    /// Initializing constructor
+    EDM4hepFileReader(const std::string& nam);
+
+    /// Read parameters set for this action
+    virtual EventReaderStatus setParameters(std::map< std::string, std::string >& parameters);
+
+    /// Read an event and fill a vector of MCParticles.
+    virtual EventReaderStatus readParticles(int event_number, Vertices& vertices, std::vector<Particle*>& particles);
+    /// Call to register the run parameters
+    void registerRunParameters();
+
+  };
+
+  /// Initializing constructor
+  dd4hep::sim::EDM4hepFileReader::EDM4hepFileReader(const std::string& nam)
+    : Geant4EventReader(nam)
+    , m_reader(podio::makeReader(nam))
+    , m_collectionName("MCParticles")
+    , m_eventHeaderCollectionName("EventHeader")
+  {
+    printout(INFO,"EDM4hepFileReader","Created file reader. Try to open input %s",nam.c_str());
+    m_directAccess = true;
+  }
+
+  void EDM4hepFileReader::registerRunParameters() {
+    try {
+      auto *parameters = new RunParameters();
+      try {
+        podio::Frame runFrame = m_reader.readFrame("runs", 0);
```

Comment:
I don't know if it's worth it here (because that would imply a podio version bump or some more pre-processor version handling), but in principle we have the `podio::Category::{Run,Event,Metadata}` constants in podio (available from 0.17 onwards) that would allow to replace some of the string constants in various places.

Reply:
Not worth it IMNSHO.